utils/fork-tree/src/lib.rs

```diff
@@ -345,18 +298,91 @@ where
 		F: Fn(&H, &H) -> Result<bool, E>,
 		P: Fn(&V) -> bool,
 	{
-		// search for node starting from all roots
-		for (index, root) in self.roots.iter().enumerate() {
-			let node = root.find_node_index_where(hash, number, is_descendent_of, predicate)?;
-
-			// found the node, early exit
-			if let FindOutcome::Found(mut node) = node {
-				node.push(index);
-				return Ok(Some(node))
+		let mut path = vec![];
+		let mut children = &self.roots;
+
+		loop {
+			let mut keep_search = false;
+			for (idx, node) in children.iter().enumerate() {
+				if node.number < *number && is_descendent_of(&node.hash, hash)? {
+					if predicate(&node.data) {
+						path.push(idx);
+						children = &node.children;
+						keep_search = true;
```

Comment:
You can get rid of `keep_search` by giving the outer loop a label:

`'main_loop: loop {`

and then doing `continue 'main_loop;` here instead of `keep_search = true`.

Reply:
<img src="https://media1.giphy.com/media/xT5LMYtyGoTGWHZZza/giphy.gif?cid=790b761133860fb18522e8ffd0c040a0ff281179c1cf2a13&rid=giphy.gif&ct=g" />

Reply:
In the end I just realised that the nested loop was not required

```
+        let mut i = 0;
+        while i < children.len() {
+            let node = &children[i];
+            if node.number < *number && is_descendent_of(&node.hash, hash)? {
+                if !predicate(&node.data) {
+                    break;
+                }
+                path.push(i);
+                i = 0;
+                children = &node.children;
+            }
+        }
```